tools/clang/lib/SPIRV/SpirvInstruction.cpp

```diff
@@ -353,153 +354,186 @@ SpirvComposite::SpirvComposite(
+SpirvConstantFloat::SpirvConstantFloat(const FloatType *type, uint16_t val)
+    : SpirvConstant(IK_ConstantFloat, spv::Op::OpConstant, type),
+      value(static_cast<uint64_t>(val)) {
+  assert(type->getBitwidth() == 16);
+}
 
-SpirvConstantFloat::SpirvConstantFloat(double val, QualType resultType,
-                                       uint32_t resultId, SourceLocation loc)
-    : SpirvConstant(IK_ConstantFloat, spv::Op::OpConstant, resultType, resultId,
-                    loc),
-      bitwidth(64), value(cast::BitwiseCast<uint64_t, double>(val)) {}
+SpirvConstantFloat::SpirvConstantFloat(const FloatType *type, float val)
+    : SpirvConstant(IK_ConstantFloat, spv::Op::OpConstant, type),
+      value(static_cast<uint64_t>(cast::BitwiseCast<uint32_t, float>(val))) {
+  assert(type->getBitwidth() == 32);
+}
+
+SpirvConstantFloat::SpirvConstantFloat(const FloatType *type, double val)
+    : SpirvConstant(IK_ConstantFloat, spv::Op::OpConstant, type),
+      value(cast::BitwiseCast<uint64_t, double>(val)) {
+  assert(type->getBitwidth() == 64);
+}
+
+uint32_t SpirvConstantFloat::getBitwidth() const {
+  // By construction, it is guaranteed spirvType to be FloatType.
+  return llvm::cast<FloatType>(resultType)->getBitwidth();
+}
 
 uint16_t SpirvConstantFloat::getValue16() const {
-  assert(bitwidth == 16);
+  assert(getBitwidth() == 16);
   return static_cast<uint16_t>(value);
 }
 
 float SpirvConstantFloat::getValue32() const {
-  assert(bitwidth == 32);
+  assert(getBitwidth() == 32);
   return cast::BitwiseCast<float, uint32_t>(static_cast<uint32_t>(value));
 }
 
 double SpirvConstantFloat::getValue64() const {
-  assert(bitwidth == 64);
+  assert(getBitwidth() == 64);
   return cast::BitwiseCast<double, uint64_t>(value);
 }
 
+bool SpirvConstantFloat::operator==(const SpirvConstantFloat &that) const {
+  return resultType == that.getResultType() && value == that.getValueBits();
+}
+
 SpirvConstantComposite::SpirvConstantComposite(
-    llvm::ArrayRef<SpirvConstant *> constituentsVec, QualType resultType,
-    uint32_t resultId, SourceLocation loc)
-    : SpirvConstant(IK_ConstantComposite, spv::Op::OpConstantComposite,
-                    resultType, resultId, loc),
-      constituents(constituentsVec) {}
-
-SpirvConstantNull::SpirvConstantNull(QualType resultType, uint32_t resultId,
-                                     SourceLocation loc)
-    : SpirvConstant(IK_ConstantNull, spv::Op::OpConstantNull, resultType,
-                    resultId, loc) {}
+    const SpirvType *type,
+    llvm::ArrayRef<const SpirvConstant *> constituentsVec)
+    : SpirvConstant(IK_ConstantComposite, spv::Op::OpConstantComposite, type),
+      constituents(constituentsVec.begin(), constituentsVec.end()) {}
+
+bool SpirvConstantComposite::
+operator==(const SpirvConstantComposite &other) const {
+  if (resultType != other.getResultType())
+    return false;
+
+  auto otherMembers = other.getConstituents();
```

Comment:
I think we can just use vector `operator==` here?

Reply:
I tried, but one side is an llvm::ArrayRef and the other is llvm::SmallVector... and apparently it can't compare those.

Reply:
Oh, I see.